From 57d1356be4f8258a99a47a4d0d48c509299fa727 Mon Sep 17 00:00:00 2001
From: Etienne Carriere <etienne.carriere@foss.st.com>
Date: Thu, 30 Nov 2023 15:11:31 +0100
Subject: [PATCH 0093/1695] plat-stm32mp1: scmi_server: remove useless bounding
 of voltage levels

Remove bounding of regulators supported voltage levels according
to mix/max levels now that drivers take care of that.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 0f3a51910e9871dc4a225fa079dd224d0b93657d plat-stm32mp1: scmi_server: remove useless bounding of voltage levels]
Change-Id: I40d32242ace114a015b92a9bf940b88b64e5ca98
Signed-off-by: Etienne Carriere <etienne.carriere@foss.st.com>
[etienne: cherry-picked from 297b2ca9531ce7bdacda4bd59d2989003b5cc331]
Signed-off-by: Etienne Carriere <etienne.carriere@foss.st.com>
---
 core/arch/arm/plat-stm32mp1/scmi_server.c | 49 +++--------------------
 1 file changed, 6 insertions(+), 43 deletions(-)

diff --git a/core/arch/arm/plat-stm32mp1/scmi_server.c b/core/arch/arm/plat-stm32mp1/scmi_server.c
index d1ab92257..b0cc34b5d 100644
--- a/core/arch/arm/plat-stm32mp1/scmi_server.c
+++ b/core/arch/arm/plat-stm32mp1/scmi_server.c
@@ -637,7 +637,6 @@ int32_t plat_scmi_voltd_levels_array(unsigned int channel_id,
 	if (voltd->regulator) {
 		struct regulator_voltages_desc *desc = NULL;
 		TEE_Result res = TEE_ERROR_GENERIC;
-		size_t ref_count = 0;
 		const int *levels = NULL;
 		size_t n = 0;
 
@@ -655,30 +654,15 @@ int32_t plat_scmi_voltd_levels_array(unsigned int channel_id,
 			return SCMI_NOT_SUPPORTED;
 		}
 
-		ref_count = desc->num_levels;
-
-		/* Bound according to regulator registered min/max levels */
-		for (n = ref_count; n > 0; n--)
-			if (levels[n - 1] > voltd->regulator->max_uv)
-				ref_count--;
-		for (n = 0; n < ref_count; n++)
-			if (levels[n] >= voltd->regulator->min_uv)
-				break;
-
-		if (n == ref_count) {
-			DMSG("Voltage list out of regulator %s level bounds",
-			     regulator_name(voltd->regulator));
-			return SCMI_GENERIC_ERROR;
-		}
-
-		if (start_index >= ref_count)
+		if (start_index >= desc->num_levels)
 			return SCMI_OUT_OF_RANGE;
 
 		if (!*nb_elts) {
-			*nb_elts = ref_count - start_index;
+			*nb_elts = desc->num_levels - start_index;
 			return SCMI_SUCCESS;
 		}
 
+		*nb_elts = MIN(*nb_elts, desc->num_levels - start_index);
 		for (n = 0; n < *nb_elts; n++)
 			out_levels[n] = levels[start_index + n];
 
@@ -700,9 +684,6 @@ int32_t plat_scmi_voltd_levels_by_step(unsigned int channel_id,
 		struct regulator_voltages_desc *desc = NULL;
 		TEE_Result res = TEE_ERROR_GENERIC;
 		const int *levels = NULL;
-		int ref_min = 0;
-		int ref_max = 0;
-		int ref_step = 0;
 
 		res = regulator_supported_voltages(voltd->regulator, &desc,
 						   &levels);
@@ -718,27 +699,9 @@ int32_t plat_scmi_voltd_levels_by_step(unsigned int channel_id,
 			return SCMI_NOT_SUPPORTED;
 		}
 
-		ref_min = levels[0];
-		ref_max = levels[1];
-		ref_step = levels[2];
-
-		if (ref_min < voltd->regulator->min_uv) {
-			int diff = voltd->regulator->min_uv - ref_min;
-			int incr = DIV_ROUND_UP(diff, ref_step);
-
-			ref_min += ref_step * incr;
-		}
-
-		if (ref_max > voltd->regulator->max_uv) {
-			int diff = ref_max - voltd->regulator->max_uv;
-			int decr = diff / ref_step;
-
-			ref_max -= ref_step * decr;
-		}
-
-		min_max_step[0] = ref_min;
-		min_max_step[1] = ref_max;
-		min_max_step[2] = ref_step;
+		min_max_step[0] = levels[0];
+		min_max_step[1] = levels[1];
+		min_max_step[2] = levels[2];
 
 		return SCMI_SUCCESS;
 	}
-- 
2.47.2

